### PR TITLE
Trunk-5119: Update apache velocity

### PR DESCRIPTION
**Pull request by gitcliff**
Updating apache velocity to 2.1
link:`https://issues.openmrs.org/browse/TRUNK-5119

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -130,8 +130,8 @@
 			</dependency>
 			<dependency>
 				<groupId>org.apache.velocity</groupId>
-				<artifactId>velocity</artifactId>
-				<version>1.6.2</version>
+				<artifactId>velocity-engine-core</artifactId>
```
Comment (HerbertYiga) [inline pom.xml]: wondering why you are changing to velocity-engine-core

File: api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java [hunk 2]
```diff
@@ -40,7 +42,15 @@ public Object methodException(Class claz, String method, Exception e) throws Exc
...
 		// keep the default behavior
-		throw e;
+		try {
+			throw e;
+		}
+		catch (Exception x) {
+			
+			x.printStackTrace();
+		}
+		return info;
```
Comment (Ruhanga) [inline api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java]: @gitcliff , what happens when you leave out this change?

Comment (gitcliff) [reply api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java]: @Ruhanga  the velocityExceptionHandler will generate a compilation error since when it throws an exception it has to be caught in the try and catch brackets

File: pom.xml (hunk 1 above)
Comment (gitcliff) [reply pom.xml]: @HerbertYiga the artifactId was changed to velocity-engine-core 2.1 
https://mvnrepository.com/artifact/org.apache.velocity/velocity-engine-core/2.1

**Comment by HerbertYiga** [comment]
@Ruhnaga do we realy need velocity engine for this? @gitcliff do we also have the  changes pass well locally?

**Comment by gitcliff** [comment]
@HerbertYiga yes they are passing well locally and they were all ticked a few days back, i dont know why the coverage/coveralls is now an error

File: api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java (hunk 2 above)
Comment (Ruhanga) [reply api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java]: @gitcliff you may want to follow java conventions such as those found here https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-MoreConventions in the case where it is necessary to print anything, but the above change is not necessary. So you may have to refactor to eliminate the 
`try {
 }
 catch {
 }`
statements.

Comment (gitcliff) [reply api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java]: @Ruhanga  i removed the try and catch blocks and refactored again

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -17,7 +17,7 @@
 import java.util.Map;
 import java.util.Set;
 
-import org.apache.commons.lang.RandomStringUtils;
+import org.apache.commons.lang3.RandomStringUtils;
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Is this required by the ticket?

File: pom.xml [hunk 4]
```diff
@@ -336,6 +336,11 @@
 				<artifactId>hapi-base</artifactId>
 				<version>2.1</version>
 			</dependency>
+			<dependency>
+				<groupId>org.codehaus.jackson</groupId>
```
Comment (dkayiwa) [inline pom.xml]: Is this a formatting change?

File: web/src/main/java/org/openmrs/web/filter/StartupFilter.java [hunk 5]
```diff
@@ -175,9 +175,9 @@ private void initializeVelocity() {
 			velocityEngine = new VelocityEngine();
 			
 			Properties props = new Properties();
-			props.setProperty(RuntimeConstants.RUNTIME_LOG, "startup_wizard_vel.log");
```
Comment (dkayiwa) [inline web/src/main/java/org/openmrs/web/filter/StartupFilter.java]: Is this required by the ticket?

File: api/src/test/java/org/openmrs/CohortTest.java [hunk 6]
```diff
@@ -9,10 +9,9 @@
  */
 package org.openmrs;
 
-import org.apache.commons.lang.ArrayUtils;
-import org.apache.commons.lang3.StringUtils;
-import org.junit.Assert;
-import org.junit.Test;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/CohortTest.java]: Are these changes required by the ticket?

File: api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java [hunk 7]
```diff
@@ -29,18 +31,19 @@
 	 * @see org.apache.velocity.app.event.MethodExceptionEventHandler#methodException(java.lang.Class,
 	 *      java.lang.String, java.lang.Exception)
 	 */
+	
 	@Override
-	public Object methodException(Class claz, String method, Exception e) throws Exception {
-		
+	public Object methodException(Context context, Class claz, String method, Exception e, Info info) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java]: Can you remind me why we needed to make this change?

File: api/pom.xml [hunk 8]
```diff
@@ -53,7 +53,7 @@
       </dependency>
       <dependency>
          <groupId>org.apache.velocity</groupId>
-         <artifactId>velocity</artifactId>
+         <artifactId>velocity-engine-core</artifactId>
```
Comment (dkayiwa) [inline api/pom.xml]: Did the artifact id change?

Comment (gitcliff) [reply api/pom.xml]: yes @dkayiwa  it changed according to this http://velocity.apache.org/download.cgi#engine

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 3 above)
Comment (gitcliff) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: no but its a dependency version that's suitable for apache velocity 2.1
http://velocity.apache.org/engine/2.1/dependencies.html

File: web/src/main/java/org/openmrs/web/filter/StartupFilter.java (hunk 5 above)
Comment (gitcliff) [reply web/src/main/java/org/openmrs/web/filter/StartupFilter.java]: yes because the new version of apache velocity 2.1 has no RUNTIME_LOG and RUNTIME_LOG_LOGSYSTEM_CLASS in its API, so i had to look for the equivalent in the new versioned API

File: pom.xml (hunk 4 above)
Comment (gitcliff) [reply pom.xml]: NO this wasn't necessary, am going to work on it

File: api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java (hunk 7 above)
Comment (gitcliff) [reply api/src/main/java/org/openmrs/util/VelocityExceptionHandler.java]: context was a parameter in the method signature of methodException method which VelocityExceptionHandler was overriding as a result of implementing the MethodExceptionEventHandle.
https://velocity.apache.org/engine/2.0/apidocs/org/apache/velocity/app/event/MethodExceptionEventHandler.html

File: api/src/test/java/org/openmrs/CohortTest.java (hunk 6 above)
Comment (gitcliff) [reply api/src/test/java/org/openmrs/CohortTest.java]: @dkayiwa these methods of Assert class where imported to be used in writing unit tests to avoid compilation errors while building

File: web/src/main/java/org/openmrs/web/filter/StartupFilter.java (hunk 5 above)
Comment (gitcliff) [reply web/src/main/java/org/openmrs/web/filter/StartupFilter.java]: https://velocity.apache.org/engine/2.0/apidocs/org/apache/velocity/runtime/RuntimeConstants.html

**Comment by teleivo** [comment]
@gitcliff could you please rebase on the latest master. Now that we are on JUnit 5 one of the tests that you touched might change.

Note: I just updated to 1.7. But since that's a minor change it should not affect your work.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 3 above)
Comment (teleivo) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:20]: if changing the UserServiceImpl is not required to do the update then please remove this change from the PR

File: api/src/test/java/org/openmrs/CohortTest.java (hunk 6 above)
Comment (teleivo) [reply api/src/test/java/org/openmrs/CohortTest.java:13]: please remove these changes from the PR

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 3 above)
Comment (gitcliff) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:20]: i made this change because` lang3.RandomStringUtils` is best suitable for `velocity 2.1` that we are upgrading to

Comment (teleivo) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:23]: ok, but how does UserServiceImpl relate to velocity?

Comment (gitcliff) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:23]: @teleivo according to  http://commons.apache.org/proper/commons-lang/ it says `Note that Lang 3.0 (and subsequent versions) use a different package (org.apache.commons.lang3) than the previous versions (org.apache.commons.lang), allowing it to be used at the same time as an earlier version` and when i build minus the above change i get this [error](https://pastebin.com/4cpbDji7)
